File: ProtoSASOnScala/src/main/scala/dd/sas/pipeline/calculation/structures/flows/associativeflows/AssociativeRefiningFlow.scala
package dd.sas.pipeline.calculation.structures.flows.associativeflows

import java.util.function.BiFunction

import dd.sas.pipeline.calculation.structures.flows.Flow
import dd.sas.pipeline.calculation.structures.nodes.Node
import dd.sas.pipeline.worldmodel.{Relation, WorldObject}

/**
  * Created by Sergey on 12.10.2016.
  */
class AssociativeRefiningFlow
[D <: WorldObject, M <: WorldObject, R <: WorldObject, S <: Relation[D, R]]
(node: Node, expression: BiFunction[D, R, S]) extends RelationFlow[D, R, S](node) {

  private var firstFlow: RelationFlow[D, M, Relation[D, M]] = _
  private var secondFlow: RelationFlow[M, R, Relation[M, R]] = _

  private var result: List[S] = _

  override def apply(): List[S] = {
    result
  }

  override def customProcess: Unit = {
    result = for {
      first <- firstFlow()
      second <- secondFlow()
      if (first.range == second.domain)
    } yield expression(first.domain, second.range)
    resubcribe()
  }


  def setDomainFlow(domainFlow: Flow): Unit = {
    this.firstFlow = domainFlow.asInstanceOf[RelationFlow[D, M, Relation[D, M]]]
    donors.put(domainFlow, this.setDomainFlow)
    setToBeUpdated()
  }

  def setRangeFlow(rangeFlow: Flow): Unit = {
    this.secondFlow = rangeFlow.asInstanceOf[RelationFlow[M, R, Relation[M, R]]]
    donors.put(rangeFlow, this.setRangeFlow)
    setToBeUpdated()
  }

}

File: ProtoSAS/src/main/java/dd/sas/pipeline/calculation/structures/flows/associativeflows/AssociativeRefiningFlow.java
package dd.sas.pipeline.calculation.structures.flows.associativeflows;

import dd.sas.pipeline.TripleFunction;
import dd.sas.pipeline.worldmodel.Relation;
import dd.sas.pipeline.worldmodel.WorldObject;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Supplier;

/**
 * Created by Sergey on 18.09.2016.
 */
public class AssociativeRefiningFlow
        <D extends WorldObject, M extends WorldObject, R extends WorldObject,
                F extends Relation<D, M>, S extends Relation<M, R>, C extends Relation<D, R>>
        extends RelationFlow<C> {

    private final TripleFunction<D, M, R, C> doStaff;
    private Supplier<List<F>> domainFlow;
    private Supplier<List<S>> rangeFlow;
    private List<C> result = new ArrayList<>();

    public AssociativeRefiningFlow(TripleFunction<D, M, R, C> doStaff) {
        this.doStaff = doStaff;
    }

    private boolean checkOnObjectEquality(M firstMiddle, M secondMiddle) {
        return firstMiddle == secondMiddle;
    }

    @Override
    public List<C> getResult() {
        return null;
    }

    @Override
    public void calculate() {
        if (checkLInFlow(domainFlow.get()) && checkLInFlow(rangeFlow.get())) {
            for (F f : domainFlow.get()) {
                for (S s : rangeFlow.get()) {
                    if (checkOnObjectEquality(f.getRange(), s.getDomain())) {
                        result.add(doStaff.apply(f.getDomain(), f.getRange(), s.getRange()));
                    }
                }
            }
        }
    }

    public void setDomainFlow(Supplier<List<F>> domainFlow) {
        this.domainFlow = domainFlow;
    }

    public void setRangeFlow(Supplier<List<S>> rangeFlow) {
        this.rangeFlow = rangeFlow;
    }


}
